File: Go/RateLimiter2-go/leaky_bucket.go
package main

import (
	"sync"
	"time"
)

type leakyBucketState struct {
	TotalRequests int
	LastLeakedAt  int64
}

type LeakyBucketStrategy struct {
	mu       sync.Mutex
	Capacity int
	LeakRate int
	Bucket   map[string]leakyBucketState
}

func NewLeakyBucketStrategy(capacity, leakRate int) *LeakyBucketStrategy {
	return &LeakyBucketStrategy{
		Capacity: capacity,
		LeakRate: leakRate,
		Bucket:   make(map[string]leakyBucketState),
	}
}

func (l *LeakyBucketStrategy) leak(ip string) {
	currentTime := time.Now().UnixMilli()
	ipRequests := l.Bucket[ip]
	timeElapsed := float64(currentTime-ipRequests.LastLeakedAt) / 1000.0
	leaks := int(float64(l.LeakRate) * timeElapsed)
	if leaks > 0 {
		l.Bucket[ip] = leakyBucketState{
			TotalRequests: max(0, ipRequests.TotalRequests-leaks),
			LastLeakedAt:  currentTime,
		}
	}
}

func (l *LeakyBucketStrategy) IsAllowed(ip string) bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.Capacity == 0 {
		return false
	}
	if _, ok := l.Bucket[ip]; !ok {
		l.Bucket[ip] = leakyBucketState{
			TotalRequests: 1,
			LastLeakedAt:  time.Now().UnixMilli(),
		}
		return true
	}
	if l.LeakRate > 0 {
		l.leak(ip)
	}
	ipRequests := l.Bucket[ip]
	if ipRequests.TotalRequests < l.Capacity {
		l.Bucket[ip] = leakyBucketState{
			TotalRequests: ipRequests.TotalRequests + 1,
			LastLeakedAt:  ipRequests.LastLeakedAt,
		}
		return true
	}
	return false
}

File: JavaScript/RateLimiter2/LeakyBucket.js
const RateLimiterStrategy = require('./RateLimiterStrategy');

class LeakyBucket extends RateLimiterStrategy {
  constructor(capacity, leakRate) {
    super();
    this.capacity = capacity;
    this.leakRate = leakRate;
    this.bucket = new Map();
  }

  leak(ip) {
    const currentTime = Date.now();
    const ipRequests = this.bucket.get(ip);

    const timeElapsed = (currentTime - ipRequests.lastLeakedAt) / 1000;
    const leaks = Math.floor(this.leakRate * timeElapsed);

    if (leaks > 0) {
      this.bucket.set(ip, {
        totalRequests: Math.max(0, ipRequests.totalRequests - leaks),
        lastLeakedAt: currentTime,
      });
    }
  }

  isAllowed(ip) {
    if (!this.capacity) return false;

    if (!this.bucket.has(ip)) {
      this.bucket.set(ip, {
        totalRequests: 1,
        lastLeakedAt: Date.now(),
      });

      return true;
    }

    if (this.leakRate) {
      this.leak(ip);
    }

    const ipRequests = this.bucket.get(ip);
    if (ipRequests.totalRequests < this.capacity) {
      this.bucket.set(ip, {
        totalRequests: ipRequests.totalRequests + 1,
        lastLeakedAt: ipRequests.lastLeakedAt,
      });
      return true;
    }

    return false;
  }
}

module.exports = LeakyBucket;
